enterprise/frontend/src/metabase-enterprise/user_provisioning/components/UserProvisioning.tsx

```diff
@@ -0,0 +1,257 @@
+import { hasAnySsoPremiumFeature } from "metabase-enterprise/settings";
+import type { EnterpriseSettings } from "metabase-enterprise/settings/types";
+import type { Settings, SettingValue } from "metabase-types/api";
+
+import { CopyScimInput, getTextInputStyles } from "./ScimInputs";
+import { ScimTextWarning } from "./ScimTextWarning";
+import {
+  UserProvisioningFirstEnabledModal,
+  UserProvisioningRegenerateTokenModal,
+} from "./UserProvisioningModals";
+
+type UserProvisioningSettings = Pick<
+  EnterpriseSettings,
+  "scim-enabled" | "scim-base-url" | "send-new-sso-user-admin-email?"
+>;
+
+export type SettingValues = Partial<Settings>;
+
+interface UserProvisioningProps {
+  elements: SettingElement[];
+  settingValues: UserProvisioningSettings &
+    Pick<
+      EnterpriseSettings,
+      | "saml-user-provisioning-enabled?"
+      | "google-auth-enabled"
+      | "ldap-enabled"
+      | "saml-enabled"
+      | "jwt-enabled"
+    >;
+  onSubmit: (values: SettingValues) => void;
+  updateSetting: (
+    settingElement: SettingElement,
+    newValue: SettingValue,
+  ) => Promise<void>;
+}
+
+// const isUnexpectedTokenError = isMatching({
+const isNoErrorOrNotFoundError = isMatching(
+  P.union(undefined, {
+    status: 404,
+  }),
+);
+
+export const UserProvisioning = ({
+  settingValues,
+  elements,
+  updateSetting,
+}: UserProvisioningProps) => {
+  const maskedTokenRequest = useGetScimTokenQuery();
+  const [regenerateToken, regenerateTokenReq] =
+    useRegenerateScimTokenMutation();
+
+  const isLoadingToken =
+    maskedTokenRequest.isLoading || regenerateTokenReq.isLoading;
+
+  const fields = useMemo(() => {
+    return _.indexBy(elements, "key");
+  }, [elements]);
+
+  const isScimEnabled = !!settingValues["scim-enabled"];
+  const isScimInitialized = !!maskedTokenRequest.data;
+
+  // since the request to enable scim and create the first token are seperate requests, it's possible
+  // that the token request failed or wasn't issued due to some interuption (network issue, instance down, etc.)
+  // we can detect this case by seing the scim is enabled without any token after all requests for token have settled
+  // this is later used to show an error message based on this token not being found + encourage the user to go through
+  // the regenerate flow to get a new unmasked token
+  const isScimIncorrectlyIniailized = Boolean(
+    !isScimInitialized &&
+      isScimEnabled &&
+      !(isLoadingToken || maskedTokenRequest.isFetching),
+  );
+
+  const samlUserProvisioningEnabled = useSetting(
+    "saml-user-provisioning-enabled?" as any,
+  );
+  const showSamlWarning = samlUserProvisioningEnabled && !isScimInitialized;
+
+  const handleScimEnabledChange = async (enabled: boolean) => {
+    const setting = fields["scim-enabled"];
+    // HACK: this is the only way to conditionally update the UI
+    // in response to the setting being successfully set
+    setting.onChanged = async () => {
```

Comment:
Lets rubber duck on this. It works, but mutating the setting feels a little gross